File: shared/src/commonMain/kotlin/org/nihongo/mochi/domain/meaning/WordMeaningRepository.kt
package org.nihongo.mochi.domain.meaning

import kotlinx.serialization.json.Json
import kotlinx.coroutines.runBlocking
import org.nihongo.mochi.domain.kana.ResourceLoader

class WordMeaningRepository(private val resourceLoader: ResourceLoader) {

    private val json = Json { 
        ignoreUnknownKeys = true 
        coerceInputValues = true
    }
    private val cachedMeanings = mutableMapOf<String, Map<String, String>>()

    fun getWordMeanings(locale: String): Map<String, String> {
        return runBlocking {
            getWordMeaningsSuspend(locale)
        }
    }

    suspend fun getWordMeaningsSuspend(locale: String): Map<String, String> {
        if (cachedMeanings.containsKey(locale)) {
            return cachedMeanings[locale]!!
        }

        val fileName = getFileName(locale)
        return try {
            val jsonString = resourceLoader.loadJson(fileName)
            if (jsonString == "{}" || jsonString.isEmpty()) throw Exception("File not found or empty: $fileName")

            val root = json.decodeFromString<WordMeaningRoot>(jsonString)
            val meaningsMap = root.word_meanings.entries.associate { it.id to it.meaning }
            cachedMeanings[locale] = meaningsMap
            meaningsMap
        } catch (e: Exception) {
            println("Error parsing word meanings for $locale (file: $fileName): ${e.message}")
            // Fallback to english (GB) if the locale is not found
            if (locale != "en_GB") {
                getWordMeaningsSuspend("en_GB")
            } else {
                emptyMap()
            }
        }
    }

    private fun getFileName(locale: String): String {
        val parts = locale.split("_")
        return if (parts.size == 2 && parts[1].length == 2 && parts[1][0].isUpperCase()) {
             "words/meanings/word_meanings_${parts[0]}_r${parts[1]}.json"
        } else {
            "words/meanings/word_meanings_${locale}.json"
        }
    }
}

File: flutter_app/lib/repositories/word_meaning_repository.dart
import 'dart:convert';
import '../services/resource_loader.dart';

class WordMeaningRepository {
  final ResourceLoader _loader;
  final Map<String, Map<String, String>> _cachedMeanings = {};

  WordMeaningRepository(this._loader);

  Future<Map<String, String>> getWordMeanings(String locale) async {
    if (_cachedMeanings.containsKey(locale)) return _cachedMeanings[locale]!;

    String effectiveLocale = locale;
    if (locale.startsWith('en')) effectiveLocale = 'en_GB';

    try {
      String jsonString = await _loader.loadString(
        'word_meanings.json',
        locale: effectiveLocale,
      );

      final decoded = json.decode(jsonString);
      List<dynamic> entries = [];
      
      if (decoded is Map) {
        final root = decoded['word_meanings'];
        if (root is List) {
          entries = root;
        } else if (root is Map) {
          final rawEntries = root['entries'] ?? root['entry'] ?? [];
          entries = rawEntries is List ? rawEntries : [rawEntries];
        }
      } else if (decoded is List) {
        entries = decoded;
      }

      final Map<String, String> meaningsMap = {};
      for (var item in entries) {
        if (item is Map) {
          final id = item['@id']?.toString() ?? item['id']?.toString();
          final meaning = item['meaning']?.toString();
          if (id != null && meaning != null) {
            meaningsMap[id] = meaning;
          }
        }
      }

      _cachedMeanings[locale] = meaningsMap;
      return meaningsMap;
    } catch (e) {
      print("Error loading word meanings for $locale: $e");
      // Fallback vers l'anglais si ce n'est pas déjà la langue demandée
      if (locale != 'en' && locale != 'en_GB') {
        return getWordMeanings('en');
      }
      return {};
    }
  }
}

